src/util/LRUCache.h

```diff
@@ -25,75 +28,141 @@ namespace ad_utility {
-        _data.pop_back();
-      }
-      assert(_data.size() <= _capacity);
+  //! Lookup a read-only value without creating a new one if non exists
+  //! instead shared_ptr<const Value>(nullptr) is returned in that case
+  shared_ptr<const Value> operator[](const Key &key) {
+    std::lock_guard<std::mutex> lock(_lock);
+    shared_ptr<const Value> result;
+    typename AccessMap::const_iterator mapIt = _accessMap.find(key);
+    if (mapIt == _accessMap.end()) {
+      // Returning a null pointer allows to easily check if the element
+      // existed and crash misuses.
+      result = shared_ptr<Value>(nullptr);
+      return result;
     }
 
-    //! Checks if there is an entry with the given key.
-    bool contains(const Key& key) {
-      return _accessMap.count(key) > 0;
+    // Move it to the front.
+    typename EntryList::iterator listIt = mapIt->second;
+    _data.splice(_data.begin(), _data, listIt);
+    _accessMap[key] = _data.begin();
+    result = _data.front().second;
+    return result;
+  }
+
+  //! Insert a key value pair to the cache.
+  void insert(const Key& key, Value value) {
+    std::lock_guard<std::mutex> lock(_lock);
+    _data.emplace_front(key, make_shared<const Value>(std::move(value)));
+    _accessMap[key] = _data.begin();
+    if (_data.size() > _capacity) {
+      // Remove the last element.
+      // Since we are using shared_ptr this does not free the underlying
+      // memory if it is still accessible through a previously returned
+      // shared_ptr
+      _accessMap.erase(_data.back().first);
+      _data.pop_back();
     }
+    assert(_data.size() <= _capacity);
+  }
 
-    //! Clear the cache
-    void clear() {
-      _data.clear();
-      _accessMap.clear();
+  //! Set the capacity.
+  void setCapacity(const size_t nofElements) {
+    std::lock_guard<std::mutex> lock(_lock);
+    _capacity = nofElements;
+    while (_data.size() > _capacity) {
+      // Remove the last element.
+      // Since we are using shared_ptr this does not free the underlying
+      // memory if it is still accessible through a previously returned
+      // shared_ptr
+      _accessMap.erase(_data.back().first);
+      _data.pop_back();
     }
+    assert(_data.size() <= _capacity);
+  }
+
+  //! Checks if there is an entry with the given key.
+  bool contains(const Key &key) {
+    std::lock_guard<std::mutex> lock(_lock);
+    return _accessMap.count(key) > 0;
+  }
+
+  //! Clear the cache
+  void clear() {
+    std::lock_guard<std::mutex> lock(_lock);
+    // Since we are using shared_ptr this does not free the underlying
+    // memory if it is still accessible through a previously returned
+    // shared_ptr
+    _data.clear();
+    _accessMap.clear();
+  }
 
-  private:
-    size_t _capacity;
-    EntryList _data;
-    AccessMap _accessMap;
+private:
+  size_t _capacity;
+  EntryList _data;
+  AccessMap _accessMap;
+  // TODO(schnelle): Once we switch to C++17
+  // this should be a shared_mutex
+  std::mutex _lock;
```

Comment:
We could use a conditional variable, that keeps track of the number of reading threads, and only notifies a waiting writing thread, once all readers have finished. The c++17 solution is definitely cleaner though.